pytorch_lightning/utilities/data.py

```diff
@@ -336,23 +352,28 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
             if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
         ]
 
-        cls = type(obj)
+        set_arg_names = []
         for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
-            if hasattr(cls, arg_name) and getattr(cls, arg_name).fset is None:
-                # the class defines a read-only (no setter) property of this name. it's likely that the implementation
-                # will set `self._arg_name = arg_value` in `__init__` which is the attribute returned by the `arg_name`
-                # property so we are fine skipping in that case
-                continue
-            setattr(obj, arg_name, arg_value)
+            if not hasattr(obj, "__" + arg_name):
+                # Set the value privately if it has not been set yet. This achieves two things:
+                # 1. The argument is the one that was passed in the outermost call
+                #    and not overriden by a call to `super().__init__()`
+                # 2. The argument is the passed value and does not get overwritten in `__init__()`
+                setattr(obj, "__" + arg_name, arg_value)
```

Comment:
(this comment applies to master too): do you think we should set a proxy here? These would be owning references otherwise, right?

Reply:
Actually, no, and it would be a bad idea to set it as a proxy. We want to be able to pull these values in the moment we are reinstantiating the instance with our parameters. And, what is even better is the fact that when we are reinstantiating, we throw away the old instance and return don't use this wrapped init, but the original one, therefore in the end, we don't have these references.